File: frontend/src/utils/roles.js
export const USER_ROLES = {
  BUYER: 'buyer',
  SELLER: 'seller',
  ADMIN: 'admin',
};

export const ROLE_LABELS = {
  [USER_ROLES.BUYER]: 'Покупатель',
  [USER_ROLES.SELLER]: 'Продавец',
  [USER_ROLES.ADMIN]: 'Администратор',
};

export function getRoleLabel(role) {
  return ROLE_LABELS[role] || 'Пользователь';
}

export function isAdmin(user) {
  if (!user) return false;
  return user.role === USER_ROLES.ADMIN || Boolean(user.is_admin) || Boolean(user.is_staff);
}

export function isSeller(user) {
  if (!user) return false;
  return user.role === USER_ROLES.SELLER || isAdmin(user);
}

export function isBuyer(user) {
  if (!user) return false;
  return [USER_ROLES.BUYER, USER_ROLES.SELLER, USER_ROLES.ADMIN].includes(user.role);
}

File: backend/api/users/roles.py
from django.db import models


class UserRole(models.TextChoices):
    BUYER = "buyer", "Покупатель"
    SELLER = "seller", "Продавец"
    ADMIN = "admin", "Администратор"


def user_has_role(user, *roles):
    if not user or not getattr(user, "is_authenticated", False):
        return False
    return getattr(user, "role", None) in roles


def is_platform_admin(user):
    if not user or not getattr(user, "is_authenticated", False):
        return False
    return (
        user_has_role(user, UserRole.ADMIN)
        or getattr(user, "is_admin", False)
        or getattr(user, "is_staff", False)
    )


def is_admin_editing_other_users_resource(user, owner_id):
    """True when a platform admin edits someone else's resource (not their own seller content)."""
    if not user or not getattr(user, "is_authenticated", False):
        return False
    return is_platform_admin(user) and owner_id != user.id


def is_seller(user):
    if not user or not getattr(user, "is_authenticated", False):
        return False
    return user_has_role(user, UserRole.SELLER, UserRole.ADMIN) or is_platform_admin(user)


def is_buyer(user):
    if not user or not getattr(user, "is_authenticated", False):
        return False
    return user_has_role(user, UserRole.BUYER, UserRole.SELLER, UserRole.ADMIN)
